### ISSUE TITLE
German Skill Store

### ISSUE THREAD

**Issue by Luftloch80** [djacobow/alexa_metar#4]
Hi can you submit the skill to german store 
or provide a zip for self hosting please

**Comment by djacobow** [comment]
I am not prepared to submit to the German store because I do not speak
German and cannot generate the appropriate utterance file. Furthermore, I
would not be in a great position to test it, either.

If you want to help you could prepare a German file that looks like:

https://github.com/djacobow/alexa_metar/blob/master/alexa_files/utterances.txt

And another that looks like:

https://github.com/djacobow/alexa_metar/blob/master/alexa_files/US_full_description.txt

I also would need to internationalize all the error prompts an other text
in the code itself. That's a bit of work, which I'm willing to take on, but I would need someone to provide the actual German strings.

The folks at Amazon are pretty picky -- they won't approve the skill unless it really works in the target language and sounds like it. They would not let me just publish an English-language skill on the DE store.

If you want to get engaged, I'd be happy to take pull requests!


It would be easier for you to host the skill yourself.  I can't just make
you a zip, though, since you would need to do a few things yourself
including:

 - creating a few dyanmodb tables
 - changing the APP_ID in the javascript to match your app
 - changing the AWS credentials loaded so that the js can access your
dynamodb tables

It's not a lot of work, but I cannot do it for you.

Best,
Dave J

**Comment by Luftloch80** [comment]
Okay i got the skill working. 
But haVe Problems with dynamo-db 
How to create them exactly please
The Other Problem is that german language is bad for reading tafs

**Comment by djacobow** [comment]
Have you been through the Alexa tutorial(s) on Amazon?

The files in the alexa_files folder in the git repo have to be copied and
pasted into the specific text boxes in the Alexa Skills Dev site. It's not
the greatest setup, but that's how Amazon does it. I would prefer it if I
could upload them directly.

I think once you've done the tutorial you will be able to see where each
goes.

Making the zip file for upload to Lambda is pretty simple. You can just use
the zip command. I do:

zip -r foo.zip *.js node_modules etc aws_config.json

I'm sure your RPi (running Raspbian?) has the zip command.


Note that if you want to self-host your own version of the app, you'll need
to change the APP_ID set at the top of index.js to match that of your skill.


Also, aws_config.json is not in the git repository, as it contains my
private credentials.

You will need your own. It looks like:

{

"accessKeyId": "xxxxxxxxxx",

"secretAccessKey": "xxxxxxxxxx",

"region": "yyyy"

}


This file is in the exact format used by the aws command line tools, so if
you just set those up and go through the prompts, it will create an
aws_config.json file for you.


Finally, you will need to use the Amazon gui to use dynamodb to set up
three tables:


weathercache (partition key "station")

forecastcache (partition key "station")

METAR_preferences (partition key "userId")


Best,

Dave J

**Comment by Luftloch80** [comment]
Thx for your help. It is now Working very Nice. But it Sounds really strange in german Language. I am an Airline Pilot and used to read Weather in englisch. But i can only Host it with german Language. It is an Amazon Problem 
Which files does the etc folder contain?

**Comment by djacobow** [comment]
Indeed, I think it is a bad choice for Amazon to make it so that people
have to use their "local" skill store. A better situation would be for
people to be able to enable a skill from *any* skill store, and
furthermore, the language and pronunciation should be what they would be
from that skill store. So, if you enable a US-based skill, you get
US-English responses with a US-English pronunciation.

That is basically how software used to work -- if there was not an
internationalized version of a program available, you could still install
that version for the whatever language it was written for, and use it.

Anyway, I'm glad the skill is working for you.

Blue skies and smooth landings,

-- dave j

**Comment by PaulDeignan** [comment]
FYI

https://developer.amazon.com/public/solutions/alexa/alexa-skills-kit/docs/developing-skills-in-multiple-languages#h2-lang-vui



Sent from my iPhone

On Feb 21, 2017, at 1:09 PM, David Jacobowitz <notifications@github.com<mailto:notifications@github.com>> wrote:

Indeed, I think it is a bad choice for Amazon to make it so that people
have to use their "local" skill store. A better situation would be for
people to be able to enable a skill from *any* skill store, and
furthermore, the language and pronunciation should be what they would be
from that skill store. So, if you enable a US-based skill, you get
US-English responses with a US-English pronunciation.

That is basically how software used to work -- if there was not an
internationalized version of a program available, you could still install
that version for the whatever language it was written for, and use it.

Anyway, I'm glad the skill is working for you.

Blue skies and smooth landings,

-- dave j

**Comment by djacobow** [comment]
I am happy to do the work to give this skill i18n support, but I will need help with german phrases for all the responses. Maybe the first step is for me to put all the strings into an i18n table and then someone who is interesting in the German could translate all of them, as well as help do the same for the utterances file. I think that would do it.

-- dave j

**Comment by Luftloch80** [comment]
I would appreciate to help but with the german skill the pronounciation is really bad. It mixes up german and english figures.

**Comment by djacobow** [comment]
I will ask on the Amazon Alexa dev support group and see how they suggest I
proceed. I would imagine the best case scenario is that all the user
utterances are in DE as well as all the non weather responses (like errors,
help, etc), but the weather should always be in English, and that might not
even be possible using the DE locale.

**Comment by Luftloch80** [comment]
Okay. I worked a little on the skill today. And i am getting closer. Now i translated most of weather to german. 
I am missing tags for TCU and CB to indicate thunderstorms. 
and it sounds better to say Wind 249 degrees with 15 knots. How to implement this?

**Comment by djacobow** [comment]
Are you making changes to a local copy of the .js?

The file that has the wind translation is wxdec_helpers.js and the function
is wind2text. You just want to change "at" to "with"?

**Comment by Luftloch80** [comment]
I am making changes to wxdec. But there is no readout for example FEW023CB Or BKN200TCU
and also CAVOK i cannot find a readout

**Comment by djacobow** [comment]
I'll add those to my code, but it looks like you're making a fork, not
preparing changes to put into this tree. Is that right?

**Comment by djacobow** [comment]
Actually, I forgot that I am not decoding the winds directly, but using the
pre-decoded information that comes from ADDS. Either they do not provide
the TCU and CB (I've never seen them before!) cloud modifiers, or I just
haven't seen a specific forecast that includes them.

If you can email me a location and time that had these, in a reasonably
timely manner (within a few hours of happening), I can look it up and see
what adds gives me. If it is there, I will add it. If it is not, I will
have to decide the cloud layers myself, which is a significant effort.

-- dave j

**Comment by djacobow** [comment]
@Luftloch80 I have added the cloud type modifiers to wxdec_helpers:

https://github.com/djacobow/alexa_metar/commit/9e80cb73779eaf6fc215c8c46ec7acda224bd928

I have not put this code on the live US/UK skill yet, as I'd like to test it a bit more first. But if you want to merge into your version, go ahead.

**Comment by Luftloch80** [comment]
thx. CB and TCU working now. Is it normal that there are no entries in man dynamodb tables?

**Comment by djacobow** [comment]
No, that's not normal. I suspect your credentials file is not present, or
not correct or the names and keys in the tables are not right.

If you look at the logs from your lambda function on cloudwatch, you should
see the error messages for the table access.

If you are hosting the skill for your own purposes only and do not plan to
put it up as a public skill, you can do without the dynamodb tables. One
table holds a user's preferences: their default airport, their default
units. If your preferences are not the default then you can just hard code
them.

The other two tables exist only to cache results from the ADDS weather
server. This is to reduce load on the server at the cost of introducing a
delay up to a few minutes on the weather data.

I added the caching tables because I thought I might get a lot of users and
did not want to get throttled back by ADDS if they saw an unreasonable
amount of traffic coming from one source. It turns out there probably are
not that many users and so the caching probably isn't necessary at all -- I
get well under 100 invocations an hour.
